Enable fast development simply by running grunt watch

almarklein opened the pull request:
issues: closes issue #2038 

When `BOKEH_RESOURCES` is in dev mode (e.g. "relative-dev" or "absolute-dev"), the JS files are taken from `bokehjs/build` instead of `bokeh/server/static`. This  means that you can simply edit the coffeescript file, grunt will generate JS, and you can refresh the page to get the new result.

bryevdv commented on bokeh/resources.py:
```diff
@@ -24,8 +25,16 @@
+
+def _static_path(path, dev=False):
+    # Prefer sources from bokehjs if dev
+    dir = _bokehjs_build_dir()
```
won't pyflakes complain about shadowing a built in? Maybe `build_dir` instead?

bryevdv commented on sphinx/source/docs/dev_guide/building.rst:
```diff
@@ -221,8 +221,9 @@ Incremental Compilation
-JavaScript modules separately. If this is done, then changes to BokehJS can be incrementally compiled, and the development iteration cycle shortened
-considerably.
+JavaScript modules separately. If this is done, then changes to BokehJS
+can be incrementally compiled (e.g. using ``grunt watch``), and the
```
Suggest: `(e.g. by running ``grunt watch`` in the ``bokehjs`` directory)`

bryevdv commented:
This generally looks good, I would like @mattpap to chime in as well.

bryevdv commented:
Look like some of the unit tests need to be updated.

mattpap commented on bokeh/resources.py:
```diff
@@ -24,8 +25,16 @@
+    dir = _bokehjs_build_dir()
+    if (not dev) or (not isdir(dir)):
+        dir = _server_static_dir()
+    # Normalize
```
Next line pretty much says this.

mattpap commented:
This should be coordinated (as much as possible) with server code in `bokeh/settings.py:85-125`. With this PR, the relevant logic is duplicated.

almarklein commented:
@mattpap I agree. As it is now, the settings and resources modules both seem to do part of the work. I'll see what I can make of it.

almarklein commented:
Interestingly, both the Resources and Settings objects have a `js_files` attribute, but the content is very different (the first has a few predefines files, the latter allmost all js files) :(  It's getting late, I'll have a look with a fresh mind tomorrow morning.

almarklein commented on bokeh/settings.py:
```diff
@@ -83,8 +84,11 @@ def simple_ids(self, default=None):
+        path = normpath(path)
+        if sys.platform == 'cygwin': path = realpath(path)
+        return path
+    
```
The normalization that we had in `resources._static_path()` is moved here.

almarklein commented:
Ok, I changed this somewhat. I modified the `bokehjsdir()` method in settings.py to accept an argument to set whether to use `bokehjs/build`. This method is then used in resources.py to get the location. 

I think it might be a good idea to at some point have a good look at these two modules (and also `server/settings.py`) to see if they can be streamlined. E.g. the server seems to use the settings object to get its resources.

bryevdv commented:
@almarklein I am not sure how I feel about this change, none of the other methods on `Settings` take any parameters besides a default, I think keeping that pattern consistent is probably valuable. 

Server settings absolutely need to be addressed, there is a different issue to make the server conform to 12-factor style configuration (env vars for most everything). So it should probably have its own `Settings` module similar to the one here, and all of the server settings should be moved there. Note at some point in the near future we are going to start packaging the server separately from the main library.

almarklein commented:
Ok, what about I move the `bokehjsdir()` function out of the class, and make `Settings.bokehjsdir()` and `Resources` call that new function.

bryevdv commented:
That seems reasonable, it could go in `bokeh.util` somewhere.

almarklein commented:
done

bryevdv commented:
great! LGTM
